quint/src/types/constraintGenerator.ts

```diff
@@ -77,9 +82,17 @@ export class ConstraintGeneratorVisitor implements IRVisitor {
     this.location = `Generating constraints for ${expressionToString(e)}`
   }
 
-  exitDef(_def: QuintDef) {
+  exitDef(def: QuintDef) {
     if (this.constraints.length > 0) {
-      this.solveConstraints()
+      this.solveConstraints().map(subs => {
+        if (!isAnnotatedDef(def)) {
+          return
+        }
```

Comment:
Actually, no. This is not `Array#map`, but [`Either#map`](https://www.npmjs.com/package/@sweet-monads/either#eithermap). It is a bit confusing, but in this case `forEach` doesn't even exist for `this.solveConstraints()` returned object.

Reply:
Ah, missed that – I should start reviewing in VSCode 😅
But since you're ignoring a `Left` returned from `solveConstraints`, can you collapse this block into the following?

```ts
    if (this.constraints.length > 0 && isAnnotatedDef(def)) {
      this.solveConstraints().map(subs => {
        checkAnnotationGenerality(subs, def.typeAnnotation).mapLeft(err =>
          this.errors.set(def.typeAnnotation?.id ?? def.id, err)
        )
      })
    }
```

Reply:
I see. `solveConstraints` does not really suggest side-effects 🙂